hail/src/main/scala/is/hail/expr/ir/EmitStream.scala

```diff
@@ -2117,105 +2165,95 @@ object EmitStream {
-                  { (_, k) =>
-                    k(
-                      COption(
-                        shuffle.getValueFinished(),
-                        EmitCode.present(
-                          shuffleType.rowDecodedPType, shuffle.getValue(eltRegion.code))))
-                  },
-                  setup = Some(EmitCodeBuilder.scopedVoid(mb) { cb =>
-                    cb.assign(shuffleLocal, CodeShuffleClient.create(
-                      mb.ecb.getType(shuffleType),
-                      uuid.loadBytes(),
-                      Code._null,
-                      mb.ecb.getPType(keyPType)))
-
-                    val startt = keyRange.loadStart(cb)
-                        .handle(cb, cb._fatal("shuffle expects defined endpoints"))
-                        .tcode[Long]
-                    val endt =
-                      keyRange.loadEnd(cb)
-                        .handle(cb, cb._fatal("shuffle expects defined endpoints"))
-                        .tcode[Long]
-
-                    cb.append(shuffle.startGet(startt, keyRange.includesStart, endt, keyRange.includesEnd))
-                  }),
-                  close = Some(Code(
-                    shuffle.getDone(),
-                    shuffle.close())))
-                unsized(stream)
-              }
-            }
-          }
+          val idt = emitIR(idIR).get(cb, "ShuffleRead cannot have null ID").asShuffle
+          val keyRangeCode =
+            emitIR(keyRangeIR).get(cb, "ShuffleRead cannot have null key range").asInterval
+
+          val uuid = idt.memoize(cb, "shuffleUUID")
+          val keyRange = keyRangeCode.memoize(cb, "shuffleClientKeyRange")
+
+          cb.ifx(!keyRange.startDefined(cb) || !keyRange.endDefined(cb), {
+            Code._fatal[Unit]("ShuffleRead cannot have null start or end points of key range")
+          })
+
+          val shuffleLocal = mb.newLocal[ShuffleClient]("shuffleClient")
+          val shuffle = new ValueShuffleClient(shuffleLocal)
+
+          val stream = (eltRegion: ChildStagedRegion) => unfold[EmitCode](
+            { (_, k) =>
+              k(
+                COption(
+                  shuffle.getValueFinished(),
+                  EmitCode.present(
+                    shuffleType.rowDecodedPType, shuffle.getValue(eltRegion.code))))
+            },
+            setup = Some(EmitCodeBuilder.scopedVoid(mb) { cb =>
+              cb.assign(shuffleLocal, CodeShuffleClient.create(
+                mb.ecb.getType(shuffleType),
+                uuid.loadBytes(),
+                Code._null,
+                mb.ecb.getPType(keyPType)))
+
+              val startt = keyRange.loadStart(cb)
+                  .get(cb, "shuffle expects defined endpoints")
+                  .asLong
+                  .tcode[Long]
+              val endt = keyRange.loadEnd(cb)
+                  .get(cb, "shuffle expects defined endpoints")
+                  .asLong
```

Comment:
these casts will fail. keyRange isn't a long, it's a struct, I think.

ShuffleClient really has the wrong interfaces and we should fix those, but for now remove the `asLong` and it should work as it did before.

Reply:
Ah, right. I added that to get it to compile, because the following `.tcode[Long]` isn't defined on `SCode`. Not clear to me the right way to fix it.

Reply:
ah! You can cast to pcode with `asPCode`

Reply:
(pcode is an scode represented by a single code, and so has tcode)

Reply:
Perfect, thanks!